Show proper error message for not allowed identifiers.

jugmac00 opened the pull request:
Due to interaction between Python(Scripts) and RestrictedPython,
following identifiers are not allowed:
- context
- container
- script
- traverse_subpath

Instead of throwing an "IndexError", now a proper error messsage
is shown.

modified:   src/Products/PythonScripts/PythonScript.py
modified:   src/Products/PythonScripts/tests/testPythonScript.py

jugmac00 committed:
Show proper error message for not allowed identifiers.

Due to interaction between Python(Scripts) and RestrictedPython,
following identifiers are not allowed:
- context
- container
- script
- traverse_subpath

Instead of throwing an "IndexError", now a proper error messsage
is shown.

This closes #33

modified: src/Products/PythonScripts/PythonScript.py
modified: src/Products/PythonScripts/tests/testPythonScript.py

jugmac00 committed:
Make flake8 happy.

modified:   src/Products/PythonScripts/PythonScript.py
modified:   src/Products/PythonScripts/tests/testPythonScript.py

dataflake reviewed (commented):
I would not define that list of identifiers in the class itself. Define it at module scope.

jugmac00 committed:
Turn class variable into module level constant.

modified:   src/Products/PythonScripts/PythonScript.py

dataflake reviewed (approved):
The PR is OK, but really, the issue is more widespread than just that. If I were to approach it I would do the following:

- a list of forbidden IDs is defined in the RestrictedPython package itself because that's where they are important. That would concentrate all "knowledge" about those identifiers in one logical central place.

- that list is checked not during object instantiation like you're doing now, but when an ObjectManager checks if an identifier is valid when placing a new object into it. That way all cases for all possible objects are covered.

Like I said, your PR is fine, I'm just describing a more general way that would immediately cover a lot more cases.

d-maurer commented:
Jens Vagelpohl wrote at 2019-7-5 05:26 -0700:
>dataflake approved this pull request.
>The PR is OK, but really, the issue is more widespread than just that. If I were to approach it I would do the following:
>
>- a list of forbidden IDs is defined in the RestrictedPython package itself because that's where they are important. That would concentrate all "knowledge" about those identifiers in one logical central place.

I do not think `RestrictedPython` is the correct place:
`RestrictedPython.compile_restricted_function` has no a-priori knowledge
about the names it should make global (and which
cause the observed error); it gets this information
from its `globalize` parameter (with values from the `PythonScript` class).


>- that list is checked not during object instantiation like you're doing now, but when an ObjectManager checks if an identifier is valid when placing a new object into it. That way all cases for all possible objects are covered.

What ids are invalid depends on the object class: you cannot use
the id `script` for a `Script (Python)`, but there is no real
reason to forbid it for e.g. a `PageTemplate`.

d-maurer commented on src/Products/PythonScripts/PythonScript.py:
```diff
@@ -61,6 +61,12 @@
+# PythonScript due to interaction with RestrictedPython.
+BAD_IDENTIFIERS = [
+    'context', 'container', 'script', 'traverse_subpath',
+]
```
Do not use an explicit list to define the problematic names, use instead the same logic as in `PythonScript._compile`: the problematic names are the `bind_names` computed at the beginning of `_compile`.

jugmac00 replied:
Thank you for pointing me to the right spot!

I did not make "bind_names" an instance variable, as it is just generated twice - or would you prefer otherwise?

d-maurer commented on src/Products/PythonScripts/PythonScript.py:
```diff
@@ -123,6 +129,11 @@ class PythonScript(Script, Cacheable):
+            raise ValueError(
+                '%s is not allowed as an identifier. '
+                'Please choose another name.' % id,
+                )
```
I would indicate here that the name has special meaning for a PythonScript and can therefore not be used as identifier. I would also indicate the other simiilar names.

jugmac00 replied:
Thank you - I updated the error message.

d-maurer commented on src/Products/PythonScripts/PythonScript.py:
```diff
@@ -123,6 +129,11 @@ class PythonScript(Script, Cacheable):
+            raise ValueError(
+                '%s is not allowed as an identifier. '
+                'Please choose another name.' % id,
+                )
```
I would indicate here that the name has a special meaning for PythonScripts and can therefore not be used as identifier. I would also name the other similar names.

jugmac00 replied:
This was already fixed in a previous commit.

jugmac00 commented:
Thank you, both, for the annotations.

I will update my PR the next time I have a spare hour (and also include a change log entry).

Thx @d-maurer for pointing me to the spot where the bind_names originate from...

jugmac00 committed:
Do not hardcode list of bad identifiers.

Also, improve error message.

modified:   src/Products/PythonScripts/PythonScript.py

d-maurer commented:
Jürgen Gmach wrote at 2019-7-6 01:56 -0700:
> ...
>I did not make "bind_names" an instance variable, as it is just generated twice - or would you prefer otherwise?

I am sad to come late with another approach - one which avoids
the need to prohibit some ids: the problematic ids fail currently
because special logic causes the defined function to arrive in
the "globals dict" while `PythonScript._newfun` looks for it in the
"locals dict". The new approach: `_newfun` looks in the "locals dict"
(as before) and if it does not find the function there, it looks in
the "globals dict" (it knows the name of the generated function).
This way, no id needs to be excluded.

We would need to check that the special names still work correctly.
But I am quite confident.

d-maurer reviewed (approved):
I approve your PR as it improves on the current state.

Unfortunately, a new idea came late to me which might be able to avoid the need to exclude some ids (details in a different comment). Up to you to decide whether you go ahead with your current PR or give the new idea a try.

jugmac00 committed:
Add change log entry.

modified:   CHANGES.rst

jugmac00 commented:
@d-maurer Thank you very much for taking your time to comment and improve the pull requests.

In a sense of "Done is better than none." I will merge this pr anyway.

If you consider to update the code with your last suggestion (and delete my changes) I am all for it and I am by no means sad/mad or whatever.

I just have no time/energy to dig deeper in the issue.

Offering a better error message is a no brainer for me, changing the behavior so that these "bad identifiers" now are allowed, is something which has to be thoroughly understood - which I don't without taking more time.
